federatedrec/optim/gradient/homo_fm_gradient.py

```diff
@@ -0,0 +1,127 @@
+#
+#  Copyright 2019 The FATE Authors. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+
+import numpy as np
+
+from arch.api.utils import log_utils
+from federatedml.util import fate_operator
+
+LOGGER = log_utils.getLogger()
+
+
+def load_data(data_instance):
+    X = []
+    Y = []
+    for iter_key, instant in data_instance:
+        weighted_feature = instant.weight * instant.features
+        X.append(weighted_feature)
+        if instant.label == 1:
+            Y.append([1])
+        else:
+            Y.append([-1])
+    X = np.array(X)
+    Y = np.array(Y)
+    return X, Y
+
+
+def fm_func(features, embed):
+    fm = []
+    feature_list = list(features)
+    for feature in feature_list:
+        re = np.multiply(np.expand_dims(feature, 1), embed)
+        # LOGGER.info("feature in fm_func:{}embed in fm_func:{}RE:{}".format(feature, embed, re))
+        re = np.sum(re, 0)
+        part1 = np.sum(np.power(re, 2))
+        features_square = np.power(feature, 2)
+        embed_square = np.power(embed, 2)
+        part2 = np.sum(np.dot(features_square, embed_square))
+        # LOGGER.info("RE after sum:{}part1:{},part2:{}".format(re, part1, part2))
+        fm.append(0.5 * (part1 - part2))
+    return fm
+
+
+class FactorizationGradient(object):
+
+    @staticmethod
+    def compute_loss(values, w, embed, intercept):
+        LOGGER.info("compute loss")
+        X, Y = load_data(values)
+        fm = fm_func(X, embed)
+        tot_loss = np.log(1 + np.exp(np.multiply(-Y.transpose(), X.dot(w) + fm + intercept))).sum()
```

Comment:
This formular can be simplified as log(1 + exp(x)), where x is something like -ywx + fm + intercept.
However, when x is relatively large like larger than 710, exp(x) will get inf which cause tot_loss inf. It is better to use log(1+exp(x)) = x if x is large enough to avoid this. 

Fate has provided a function in optim/activation called log_logistic. log(1+exp(x)) can safely be calculated as -log_logistic(-x) to solve this problem.